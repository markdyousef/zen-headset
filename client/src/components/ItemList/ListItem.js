import React from "react";
import {
  ExpansionPanel,
  ExpansionPanelActions,
  ExpansionPanelSummary,
  Chip,
  CircularProgress
} from "@material-ui/core";
import Divider from "@material-ui/core/Divider";
import Button from "@material-ui/core/Button";
import ExpandMoreIcon from "@material-ui/icons/ExpandMore";
import moment from "moment";
import styled from "styled-components";

const LoaderContainer = styled.section`
  z-index: 999;
  height: 150px;
  width: 100%;
  position: absolute;
  background: rgba(255, 255, 255, 0.9);
  top: 0;
  display: flex;
  align-items: center;
  justify-content: center;
`;

const Loader = () => <LoaderContainer>
  <CircularProgress />
</LoaderContainer>;
/**
 * Expansion Panel Styles
 */
const Header = styled.header`
  width: 100%;
  display: flex;
  align-items: baseline;
  position: relative;
  justify-content: space-between;
  h3 {
    font-size: 14px;
    margin: 5px;
    font-weight: normal;
  }
  h5 {
    font-size: 12px;
    margin: 5px;
    font-weight: bold;
  }
`;

const ScoreContainer = styled.span`
  color: #ddd;
  margin-right: 10px;
  max-width: 50px;
`;

const Score = ({ score }) => (
  <ScoreContainer>
    <Chip label={score} />
  </ScoreContainer>
);

const Title = styled.div`
  width: 100%;
`;

const Time = styled.span`
  min-width: 100px;
`;

const Link = styled.a`
  text-decoration: none;
  color: inherit;
`;

const Panel = styled(ExpansionPanel)`
  position: relative;
`;

export const ListItem = ({
  item,
  handleChange,
  expanded,
  handleSave,
  isLoading
}) => {
  const time = moment.unix(item.time).fromNow();
  return (
    <Panel expanded={expanded} onChange={handleChange}>
      <ExpansionPanelSummary expandIcon={<ExpandMoreIcon />}>
        <Header>
          <Score score={item.score} />
          <Title>
            <h3>
              <Link href={item.url} target="_blank">
                {item.title}
              </Link>
            </h3>
            <h5>@{item.by}</h5>
          </Title>
          <Time>{time}</Time>
        </Header>
      </ExpansionPanelSummary>
      <Divider />
      <ExpansionPanelActions>
        <Button size="small" color="primary" onClick={handleSave}>
          Save
        </Button>
        <Button size="small">Close</Button>
      </ExpansionPanelActions>
      {isLoading && <Loader />}
    </Panel>
  );
};

export default ListItem;
