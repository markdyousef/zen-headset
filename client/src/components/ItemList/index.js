import React from "react";
import ListItem from "./ListItem";
import styled from "styled-components";
import { CircularLoader } from "../common";
import { Snackbar } from "@material-ui/core";

const ListContainer = styled.section`
  margin-top: 2px;
`;

class ItemList extends React.Component {
  componentWillMount() {
    this.props.getItems();
  }
  renderList = () => {
    const { items, isLoading, activeItem, saveItem, expandItem } = this.props;
    if (isLoading) return <CircularLoader />;
    return items.map(item => (
      <ListItem
        item={item}
        key={item.id}
        handleChange={() => expandItem(item.id)}
        expanded={item.id === this.props.expandedItem}
        handleSave={() => saveItem(item)}
        isLoading={
          item.id == this.props.expandedItem ? activeItem.isLoading : false
        }
      />
    ));
  };
  renderMessages = () => {
    const { activeItem } = this.props;

    // item is saved!
    if (!activeItem.isLoading && activeItem.isSaved) {
      // setTimeout(() => ({ showMessage: false }), 1000);
      return (
        <Snackbar
          anchorOrigin={{ vertical: "bottom", horizontal: "left" }}
          open={activeItem.message}
          message={<span>{activeItem.message}</span>}
        />
      );
    }
  };
  render() {
    return (
      <ListContainer>
        {this.renderList()}
        {this.renderMessages()}
      </ListContainer>
    );
  }
}

export default ItemList;
