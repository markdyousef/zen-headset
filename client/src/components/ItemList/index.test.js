import React from "react";
import ItemList from "./index";
import { shallow } from "enzyme";

const ITEMS = [
  {
    by: "rchaudhary",
    descendants: 9,
    id: 17256923,
    kids: [17259028, 17258582, 17258792, 17258618, 17259069, 17258445],
    score: 65,
    time: 1528386584,
    title: "Raising a Very Unusual Round of Funding and Open-Sourcing Our Docs",
    type: "story",
    url:
      "https://sparktoro.com/blog/raised-a-very-unusual-round-of-funding-were-open-sourcing-our-docs/"
  }
];

describe("ItemList", () => {
  test("should match snapshot", () => {
    const wrapper = shallow(
      <ItemList items={ITEMS} getItems={() => {}} activeItem={{}} />
    );
    expect(wrapper).toMatchSnapshot();
  });
});
