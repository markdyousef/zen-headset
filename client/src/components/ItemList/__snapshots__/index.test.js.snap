// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ItemList should match snapshot 1`] = `
ShallowWrapper {
  "length": 1,
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <ItemList
    items={
        Array [
            Object {
              "by": "rchaudhary",
              "descendants": 9,
              "id": 17256923,
              "kids": Array [
                17259028,
                17258582,
                17258792,
                17258618,
                17259069,
                17258445,
              ],
              "score": 65,
              "time": 1528386584,
              "title": "Raising a Very Unusual Round of Funding and Open-Sourcing Our Docs",
              "type": "story",
              "url": "https://sparktoro.com/blog/raised-a-very-unusual-round-of-funding-were-open-sourcing-our-docs/",
            },
          ]
    }
/>,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "host",
    "props": Object {
      "children": Array [
        Array [
          <Unknown
            expanded={false}
            handleChange={[Function]}
            handleSave={[Function]}
            item={
                        Object {
                                    "by": "rchaudhary",
                                    "descendants": 9,
                                    "id": 17256923,
                                    "kids": Array [
                                      17259028,
                                      17258582,
                                      17258792,
                                      17258618,
                                      17259069,
                                      17258445,
                                    ],
                                    "score": 65,
                                    "time": 1528386584,
                                    "title": "Raising a Very Unusual Round of Funding and Open-Sourcing Our Docs",
                                    "type": "story",
                                    "url": "https://sparktoro.com/blog/raised-a-very-unusual-round-of-funding-were-open-sourcing-our-docs/",
                                  }
            }
            openDetail={[Function]}
/>,
        ],
        false,
      ],
    },
    "ref": null,
    "rendered": Array [
      Object {
        "instance": null,
        "key": "17256923",
        "nodeType": "function",
        "props": Object {
          "expanded": false,
          "handleChange": [Function],
          "handleSave": [Function],
          "item": Object {
            "by": "rchaudhary",
            "descendants": 9,
            "id": 17256923,
            "kids": Array [
              17259028,
              17258582,
              17258792,
              17258618,
              17259069,
              17258445,
            ],
            "score": 65,
            "time": 1528386584,
            "title": "Raising a Very Unusual Round of Funding and Open-Sourcing Our Docs",
            "type": "story",
            "url": "https://sparktoro.com/blog/raised-a-very-unusual-round-of-funding-were-open-sourcing-our-docs/",
          },
          "openDetail": [Function],
        },
        "ref": null,
        "rendered": null,
        "type": [Function],
      },
      false,
    ],
    "type": "div",
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "host",
      "props": Object {
        "children": Array [
          Array [
            <Unknown
              expanded={false}
              handleChange={[Function]}
              handleSave={[Function]}
              item={
                            Object {
                                          "by": "rchaudhary",
                                          "descendants": 9,
                                          "id": 17256923,
                                          "kids": Array [
                                            17259028,
                                            17258582,
                                            17258792,
                                            17258618,
                                            17259069,
                                            17258445,
                                          ],
                                          "score": 65,
                                          "time": 1528386584,
                                          "title": "Raising a Very Unusual Round of Funding and Open-Sourcing Our Docs",
                                          "type": "story",
                                          "url": "https://sparktoro.com/blog/raised-a-very-unusual-round-of-funding-were-open-sourcing-our-docs/",
                                        }
              }
              openDetail={[Function]}
/>,
          ],
          false,
        ],
      },
      "ref": null,
      "rendered": Array [
        Object {
          "instance": null,
          "key": "17256923",
          "nodeType": "function",
          "props": Object {
            "expanded": false,
            "handleChange": [Function],
            "handleSave": [Function],
            "item": Object {
              "by": "rchaudhary",
              "descendants": 9,
              "id": 17256923,
              "kids": Array [
                17259028,
                17258582,
                17258792,
                17258618,
                17259069,
                17258445,
              ],
              "score": 65,
              "time": 1528386584,
              "title": "Raising a Very Unusual Round of Funding and Open-Sourcing Our Docs",
              "type": "story",
              "url": "https://sparktoro.com/blog/raised-a-very-unusual-round-of-funding-were-open-sourcing-our-docs/",
            },
            "openDetail": [Function],
          },
          "ref": null,
          "rendered": null,
          "type": [Function],
        },
        false,
      ],
      "type": "div",
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
      },
    },
  },
}
`;
